<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>GRIMER Reports - Examples - GRIMER</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "GRIMER Reports - Examples";
        var mkdocs_page_input_path = "examples.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> GRIMER
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">GRIMER</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../importing/">Importing files</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../config/">Configuration file</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../manual/">GRIMER Reports - User Manual</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">GRIMER Reports - Examples</a>
    <ul class="current">
    </ul>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">GRIMER</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li>
      <li>GRIMER Reports - Examples</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <p><img alt="GRIMER" src="https://raw.githubusercontent.com/pirovc/grimer/main/grimer/img/logo.png" /></p>
<p>Examples of reports generated with <a href="https://github.com/pirovc/grimer">GRIMER</a></p>
<hr />
<h3 id="data-analysis-from-leiby-et-al-lack-of-detection-of-a-human-placenta-microbiome-in-samples-from-preterm-and-term-deliveries">Data analysis from Leiby et al. "Lack of detection of a human placenta microbiome in samples from preterm and term deliveries"</h3>
<p><strong><em>original publication: <a href="https://doi.org/10.1186/s40168-018-0575-4">10.1186/s40168-018-0575-4</a></em></strong></p>
<p><strong><a href="https://pirovc.github.io/grimer-reports/placenta/placenta_mgs.html">GRIMER report MGS</a></strong></p>
<p><strong><a href="https://pirovc.github.io/grimer-reports/placenta/placenta_amplicon.html">GRIMER report AMPLICON</a></strong></p>
<details>
<summary>commands used to create report</summary>


<pre><code class="language-bash"># Download files (table, metadata and config)
wget https://raw.githubusercontent.com/pirovc/grimer-reports/main/placenta/placenta_files.tar.gz
tar xf  placenta_files.tar.gz

# AMPLICON
grimer --config placenta_amplicon_config.yaml \
       --input-file placenta_amplicon_table.tsv \
       --metadata-file placenta_metadata.tsv \
       --taxonomy ncbi \
       --ranks superkingdom phylum class order family genus species \
       --level-separator &quot;;&quot; \
       --obs-replace &quot;^.+__&quot; &quot;&quot; &quot;_&quot; &quot; &quot; \
       --unassigned-header &quot;Unassigned&quot;  \
       --decontam --mgnify --transpose \
       --title &quot;Placenta study AMPLICON - Leiby, J.S. et al 2018&quot; \
       --output-html placenta_amplicon.html

# MGS
grimer --config placenta_mgs_config.yaml \
       --input-file placenta_mgs_table.tsv \
       --metadata-file placenta_metadata.tsv \
       --taxonomy ncbi \
       --ranks superkingdom phylum class order family genus species \
       --level-separator &quot;|&quot; \
       --unassigned-header &quot;unassigned&quot;  \
       --decontam --mgnify \
       --title &quot;Placenta study MGS - Leiby, J.S. et al 2018&quot; \
       --output-html placenta_mgs.html
</code></pre>


</details>
<hr />
<h3 id="katharoseq-analysis-from-minich-et-al-katharoseq-enables-high-throughput-microbiome-analysis-from-low-biomass-samples">KatharoSeq analysis from Minich et al. "KatharoSeq Enables High-Throughput Microbiome Analysis from Low-Biomass Samples"</h3>
<p><strong><em>original publication: <a href="https://doi.org/10.1128/mSystems.00218-17">10.1128/mSystems.00218-17</a></em></strong></p>
<p><strong><a href="https://pirovc.github.io/grimer-reports/katharoseq/katharoseq.html">GRIMER report</a></strong></p>
<details>
<summary>commands used to create report</summary>


<pre><code class="language-bash"># Download files (table, metadata and config)
wget https://raw.githubusercontent.com/pirovc/grimer-reports/main/katharoseq/katharoseq_files.tar.gz
tar xf katharoseq_files.tar.gz

# Run GRIMER
grimer --config katharoseq_config.yaml \
       --input-file katharoseq_table.tsv \
       --metadata-file katharoseq_metadata.tsv \
       --transformation clr \
       --obs-replace &quot;^.+__&quot; &quot;&quot; &quot;_&quot; &quot; &quot; \
       --taxonomy ncbi \
       --ranks superkingdom phylum class order family genus species \
       --level-separator &quot;;&quot; \
       --decontam --mgnify \
       --title &quot;KatharoSeq - Minich et al. 2018&quot; \
       --output-html katharoseq.html
</code></pre>


</details>

<hr />
<h3 id="preterm-infant-resistome-downloaded-from-microbiomedb">Preterm Infant Resistome downloaded from <a href="https://microbiomedb.org/mbio/app/record/dataset/DS_82fe0308e2">MicrobiomeDB</a></h3>
<p><strong><em>original publication: <a href="https://doi.org/10.1038/nmicrobiol.2016.24">10.1038/nmicrobiol.2016.24</a></em></strong></p>
<p><strong><a href="https://pirovc.github.io/grimer-reports/microbiomedb/ResistomeAmplicon.html">GRIMER report</a></strong></p>
<details>
<summary>commands used to create report</summary>


<pre><code class="language-bash"># Download files (table, metadata and config) - Original source: https://microbiomedb.org/common/downloads/release-22/82fe0308e2032de2041694df6592ba542ea84b86/ResistomeAmplicon.16s_DADA2.taxon_abundance.biom
wget https://raw.githubusercontent.com/pirovc/grimer-reports/main/microbiomedb/microbiomedb_files.tar.gz
tar xf microbiomedb_files.tar.gz

# Run GRIMER
grimer --config config/default.yaml \
       --input-file ResistomeAmplicon.16s_DADA2.taxon_abundance.biom \
       --taxonomy ncbi \
       --ranks superkingdom phylum class order family genus species \
       --decontam --mgnify \
       --title &quot;MicrobiomeDB Preterm Infant Resistome (V4)&quot; \
       --output-html ResistomeAmplicon.html
</code></pre>


</details>

<hr />
<h3 id="antibiotic-induced-changes-in-the-microbiota-disrupt-redox-dynamics-in-the-gut-downloaded-from-mgnify">Antibiotic induced changes in the microbiota disrupt redox dynamics in the gut downloaded from <a href="https://www.ebi.ac.uk/metagenomics/studies/MGYS00005180">MGnify</a></h3>
<p><strong><em>original publication <a href="https://doi.org/10.7554/elife.35987">10.7554/elife.35987</a></em></strong></p>
<p><strong><a href="https://pirovc.github.io/grimer-reports/mgnify/MGYS00005180.html">GRIMER report</a></strong></p>
<details>
<summary>commands used to create report</summary>


<pre><code class="language-bash"># Script to download files and generate GRIMER report from any MGnify study accession
# Requires &quot;jsonapi-client&gt;=0.9.7&quot; (conda install &quot;jsonapi-client&gt;=0.9.7&quot;)
./grimer-mgnify.py -i MGYS00005180 -o MGYS00005180 -g &quot;--decontam --mgnify&quot; 

# Or directly from files
wget https://raw.githubusercontent.com/pirovc/grimer-reports/main/mgnify/mgnify_files.tar.gz
tar xf mgnify.tar.gz
# Run GRIMER
grimer --config config/default.yaml \
       --input-file MGYS00005180_ERP108433_taxonomy_abundances_SSU_v4.1.tsv \
       --metadata-file MGYS00005180_metadata.tsv \
       --obs-replace &quot;^.+__&quot; &quot;&quot; &quot;_&quot; &quot; &quot; \
       --taxonomy ncbi \
       --ranks superkingdom kingdom phylum class order family genus species \
       --level-separator &quot;;&quot; \
       --decontam --mgnify \
       --title &quot;MGnify study accession MGYS00005180&quot; \
       --output-html MGYS00005180.html
</code></pre>


</details>

<hr />
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../manual/" class="btn btn-neutral float-left" title="GRIMER Reports - User Manual"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../manual/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
